services/blockstorage/sync/block_sync.go

```diff
@@ -0,0 +1,121 @@
+	"github.com/orbs-network/orbs-spec/types/go/services/gossiptopics"
+	"sync"
+	"time"
+)
+
+type syncState interface {
+	name() string
+	processState(ctx context.Context) syncState
+	blockCommitted()
+	gotAvailabilityResponse(message *gossipmessages.BlockAvailabilityResponseMessage)
+	gotBlocks(message *gossipmessages.BlockSyncResponseMessage)
+}
+
+type blockSyncConfig interface {
+	NodePublicKey() primitives.Ed25519PublicKey
+	BlockSyncBatchSize() uint32
+	BlockSyncNoCommitInterval() time.Duration
+	BlockSyncCollectResponseTimeout() time.Duration
+	BlockSyncCollectChunksTimeout() time.Duration
+}
+
+type BlockSyncStorage interface {
+	GetBlocks(first primitives.BlockHeight, last primitives.BlockHeight) (blocks []*protocol.BlockPairContainer, firstAvailableBlockHeight primitives.BlockHeight, lastAvailableBlockHeight primitives.BlockHeight)
+	LastCommittedBlockHeight() primitives.BlockHeight
+	CommitBlock(input *services.CommitBlockInput) (*services.CommitBlockOutput, error)
+	ValidateBlockForCommit(input *services.ValidateBlockForCommitInput) (*services.ValidateBlockForCommitOutput, error)
+	UpdateConsensusAlgosAboutLatestCommittedBlock()
+}
+
+type BlockSync struct {
+	logger       log.BasicLogger
+	terminated   bool
+	sf           *stateFactory
+	gossip       gossiptopics.BlockSync
+	storage      BlockSyncStorage
+	config       blockSyncConfig
+	currentState syncState
+	eventLock    *sync.Mutex
+}
+
+func NewBlockSync(ctx context.Context, config blockSyncConfig, gossip gossiptopics.BlockSync, storage BlockSyncStorage, logger log.BasicLogger) *BlockSync {
+	bs := &BlockSync{
+		logger:     logger.WithTags(log.String("flow", "block-sync")),
+		terminated: false,
+		sf:         NewStateFactory(config, gossip, storage, logger),
+		gossip:     gossip,
+		storage:    storage,
+		config:     config,
+		eventLock:  &sync.Mutex{},
+	}
+
+	bs.logger.Info("block sync init",
+		log.Stringable("no-commit-timeout", bs.config.BlockSyncNoCommitInterval()),
+		log.Stringable("collect-responses-timeout", bs.config.BlockSyncCollectResponseTimeout()),
+		log.Stringable("collect-chunks-timeout", bs.config.BlockSyncCollectChunksTimeout()),
+		log.Uint32("batch-size", bs.config.BlockSyncBatchSize()))
+
+	go bs.syncLoop(ctx)
+	return bs
+}
+
+func (bs *BlockSync) syncLoop(ctx context.Context) {
+	meter := bs.logger.Meter("inter-sync-main-loop")
+	for bs.currentState = bs.sf.CreateIdleState(); bs.currentState != nil; {
+		bs.logger.Info("state transitioning", log.String("current-state", bs.currentState.name()))
+		bs.currentState = bs.currentState.processState(ctx)
+	}
+
+	bs.terminated = true
+	meter.Done()
+}
+
+func (bs *BlockSync) HandleBlockCommitted() {
+	bs.eventLock.Lock()
+	defer bs.eventLock.Unlock()
+	if bs.currentState != nil {
+		bs.currentState.blockCommitted()
+	}
+}
+
+func (bs *BlockSync) HandleBlockAvailabilityRequest(input *gossiptopics.BlockAvailabilityRequestInput) (*gossiptopics.EmptyOutput, error) {
```

Comment:
so, again - the *service* implements the handler - since it must do that as per protos. It then delegates the *client* stuff to the state machine and the *server* stuff to another unit, but they don't have to implement the handler interface.

Reply:
yeah, i agree, will be fixed, i will keep the code under the sync package still, but the service will implement the handlers and delegate accordingly (not sure as part of this pr)